plugin/helpers_bots.go

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package plugin
+
+import (
+	"time"
+
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/pkg/errors"
+)
+
+func (p *HelpersImpl) EnsureBot(bot *model.Bot) (retBotId string, retErr error) {
+	// Must provide a bot with a username
+	if bot == nil || len(bot.Username) < 1 {
+		return "", errors.New("EnsureBot was passed a bad bot")
+	}
+
+	// If we fail for any reason, this could be a race between creation of bot and
+	// retreval from anouther EnsureBot. Just try the basic retrieve existing again.
+	defer func() {
+		if retBotId == "" || retErr != nil {
+			time.Sleep(time.Second)
+			botIdBytes, err := p.API.KVGet(BOT_USER_KEY)
+			if err == nil && botIdBytes != nil {
+				retBotId = string(botIdBytes)
+				retErr = nil
+			}
+		}
+	}()
+
+	botIdBytes, kvGetErr := p.API.KVGet(BOT_USER_KEY)
+	if kvGetErr != nil {
+		return "", errors.Wrap(kvGetErr, "Failed to get bot in EnsureBot")
```

Comment:
Nit: it's idiomatic that errors shouldn't start with a capital latter and I would like to stick with this, as this is a new library.
```suggestion
		return "", errors.Wrap(kvGetErr, "failed to get bot")
```